MembershipFailureTest_withTCP slave_receives_member_list_from_non_master

- Fails on `Hazelcast-4.master-nightly`
- Fails on [Build #539 (Jul 17, 2020 5:07:00 PM)](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.master-nightly/539/testReport/junit/com.hazelcast.internal.cluster.impl/MembershipFailureTest_withTCP/slave_receives_member_list_from_non_master_fd_deadline_/)
- Error
```
Cluster size is not correct. Expected: 2, actual: 3, instance index: 0
```
- Stacktrace
```
java.lang.AssertionError: Cluster size is not correct. Expected: 2, actual: 3, instance index: 0
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSize(HazelcastTestSupport.java:1029)
	at com.hazelcast.test.HazelcastTestSupport$11.run(HazelcastTestSupport.java:1057)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1242)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1259)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1054)
	at com.hazelcast.test.HazelcastTestSupport.assertClusterSizeEventually(HazelcastTestSupport.java:1042)
	at com.hazelcast.internal.cluster.impl.MembershipFailureTest.slave_receives_member_list_from_non_master(MembershipFailureTest.java:361)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
```
- It could manifest since hiccups, as also `FailOnTimeoutStatement` in stacktrace.
```
18:09:35, accumulated pauses: 35 ms, max pause: 1 ms, pauses over 1000 ms: 0
18:09:40, accumulated pauses: 60 ms, max pause: 4 ms, pauses over 1000 ms: 0
18:09:45, accumulated pauses: 49 ms, max pause: 6 ms, pauses over 1000 ms: 0
18:09:50, accumulated pauses: 47 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:09:55, accumulated pauses: 48 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:00, accumulated pauses: 45 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:05, accumulated pauses: 31 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:10, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:15, accumulated pauses: 30 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:20, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:25, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:30, accumulated pauses: 31 ms, max pause: 2 ms, pauses over 1000 ms: 0
18:10:35, accumulated pauses: 29 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:40, accumulated pauses: 31 ms, max pause: 1 ms, pauses over 1000 ms: 0
18:10:45, accumulated pauses: 32 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:50, accumulated pauses: 35 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:10:55, accumulated pauses: 42 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:11:00, accumulated pauses: 58 ms, max pause: 1 ms, pauses over 1000 ms: 0
18:11:05, accumulated pauses: 45 ms, max pause: 0 ms, pauses over 1000 ms: 0
18:11:10, accumulated pauses: 44 ms, max pause: 3 ms, pauses over 1000 ms: 0
18:11:15, accumulated pauses: 51 ms, max pause: 2 ms, pauses over 1000 ms: 0
18:11:20, accumulated pauses: 49 ms, max pause: 3 ms, pauses over 1000 ms: 0
18:11:25, accumulated pauses: 49 ms, max pause: 1 ms, pauses over 1000 ms: 0
18:11:30, accumulated pauses: 45 ms, max pause: 3 ms, pauses over 1000 ms: 0
18:11:35, accumulated pauses: 50 ms, max pause: 4 ms, pauses over 1000 ms: 0
18:11:40, accumulated pauses: 71 ms, max pause: 2 ms, pauses over 1000 ms: 0
18:11:45, accumulated pauses: 315 ms, max pause: 285 ms, pauses over 1000 ms: 0
```
- kindly check

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnection.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnection.java
@@ -318,6 +318,7 @@ public class TcpServerConnection implements ServerConnection {
                 + ", endpoint=" + remoteAddress
                 + ", alive=" + alive
                 + ", connectionType=" + connectionType
+                + ", planeIndex=" + planeIndex
                 + "]";
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager.java
@@ -45,6 +45,7 @@ import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import java.io.IOException;
 import java.nio.channels.SocketChannel;
 import java.util.Collection;
+import java.util.Iterator;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
@@ -523,7 +524,27 @@ public class TcpServerConnectionManager
                     plane.connectionsInProgress.remove(remoteAddress);
                     plane.connectionMap.remove(remoteAddress);
                     fireConnectionRemovedEvent(connection, remoteAddress);
-                } //todo: could it be that we have a memory leak by not removing something from the plane.
+                } else {
+                    // it might be the case that the connection was closed quickly enough
+                    // that the planeIndex was not set. Instead look for the connection
+                    // by remoteAddress and connectionId over all planes and remove it wherever found.
+                    boolean removed = false;
+                    for (Plane plane : planes) {
+                        plane.connectionsInProgress.remove(remoteAddress);
+                        // not using removeIf due to https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8078645
+                        Iterator<TcpServerConnection> connections = plane.connectionMap.values().iterator();
+                        while (connections.hasNext()) {
+                            TcpServerConnection cxn = connections.next();
+                            if (cxn.getConnectionId() == connection.getConnectionId()) {
+                                connections.remove();
+                                removed = true;
+                            }
+                        }
+                    }
+                    if (removed) {
+                        fireConnectionRemovedEvent(connection, remoteAddress);
+                    }
+                }
             }
 
             if (cause != null) {

--- a/hazelcast/src/test/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager_MemoryLeakTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/server/tcp/TcpServerConnectionManager_MemoryLeakTest.java
@@ -29,15 +29,8 @@ import org.junit.runner.RunWith;
 
 import static com.hazelcast.test.Accessors.getNode;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
 
-/**
- * THis is a test for https://github.com/hazelcast/hazelcast-enterprise/issues/2492
- * The cause of the problem is the new pipeline in 3.11.
- * In the old approach, the channel wasn't registered before the connection was established.
- * So in case of failure, nothing needs to be unregistered.
- * But with the new pipeline the channel gets registered (created) before the connection is established
- * but it didn't get unregistered if the connection could not be established. Leading to a memory leak.
- */
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(QuickTest.class)
 public class TcpServerConnectionManager_MemoryLeakTest
@@ -48,6 +41,14 @@ public class TcpServerConnectionManager_MemoryLeakTest
         Hazelcast.shutdownAll();
     }
 
+    /**
+     * This is a test for https://github.com/hazelcast/hazelcast-enterprise/issues/2492
+     * The cause of the problem is the new pipeline in 3.11.
+     * In the old approach, the channel wasn't registered before the connection was established.
+     * So in case of failure, nothing needs to be unregistered.
+     * But with the new pipeline the channel gets registered (created) before the connection is established
+     * but it didn't get unregistered if the connection could not be established. Leading to a memory leak.
+     */
     @Test
     public void test() {
         HazelcastInstance hz1 = Hazelcast.newHazelcastInstance();
@@ -61,5 +62,30 @@ public class TcpServerConnectionManager_MemoryLeakTest
         TcpServerConnectionManager connectionManager = networkingService.getConnectionManager(EndpointQualifier.MEMBER);
 
         assertTrueAllTheTime(() -> assertEquals(0, connectionManager.acceptedChannels.size()), 5);
+    }
+
+    /**
+     * When a connection is quickly closed by the remote side, it can be the case
+     * that the {@link com.hazelcast.internal.cluster.impl.MemberHandshake}
+     * was not processed on the originating member and the
+     * {@code TcpServerConnection#planeIndex} was left to its default value (-1).
+     * This race results in further connection attempts being blocked due to
+     * leftover data in {@code Plane#connectionsInProgress}.
+     * See https://github.com/hazelcast/hazelcast/issues/17238#issuecomment-694241589
+    */
+    @Test
+    public void test_connectionCleanedUpWhenClosed_withoutPlaneIndexSet() {
+        HazelcastInstance hz1 = Hazelcast.newHazelcastInstance();
+        TcpServer networkingService = (TcpServer) getNode(hz1).getServer();
+
+        HazelcastInstance hz2 = Hazelcast.newHazelcastInstance();
+        assertClusterSizeEventually(2, hz1);
+        TcpServerConnectionManager connectionManager = networkingService.getConnectionManager(EndpointQualifier.MEMBER);
+        TcpServerConnection connection = (TcpServerConnection) connectionManager.get(getNode(hz2).getThisAddress());
+        connection.setPlaneIndex(-1);
+        hz2.shutdown();
+
+        assertClusterSizeEventually(1, hz1);
+        assertTrue(connectionManager.getConnections().isEmpty());
     }
 }

